### ISSUE TITLE
[Bug]: KeyError raised when Using GroupKFold for Regression Tasks

### ISSUE THREAD

**Issue by dannycg1996** [microsoft/FLAML#1384]
### Describe the bug

Hi,

I've found an issue where group folds cannot be used for regression tasks. There are (as far as I know) four requirements for this error to occur.

1. The task is a regression task.
2. We have `split_type='group'`.
3. `y_train` (passed to `AutoML.fit()` is a Pandas Series.
4. The indices of `y_train` aren't equal to `range(len(y_train.shape[0]))` i.e. some indices may be out of order/missing.

Essentially when these criteria are met, we receive a KeyError, such as the one shown below:

![Image](https://github.com/user-attachments/assets/9019b74a-6cdf-49f7-8a4f-67fa2cc7b193)

It's worth mentioning that if we don't use GroupKFolds (i.e. we set `split_type='uniform'`), then this issue doesn't occur - everything works well.

### Steps to reproduce

An example piece of code which generates this error is shown below. Essentially we use the Iris dataset as a regression dataset, and append a nonsensical cluster column to the dataset. We then split this data into a train and test set. Here we are using FLAML to perform the train and validation steps, whilst we withhold the test set to measure performance against afterwards.

```
from sklearn import datasets
import numpy as np
from flaml import AutoML
from sklearn.model_selection import train_test_split

iris_dict_data = datasets.load_iris(as_frame=True)  # numpy arrays
iris_data = iris_dict_data["frame"]  # pandas dataframe data + target
iris_data['cluster'] = np.random.randint(0, 5, iris_data.shape[0])
automl = AutoML()

X= iris_data[["sepal length (cm)","sepal width (cm)", "petal length (cm)"]].to_numpy()
y = iris_data["petal width (cm)"]
X_train, X_test, y_train, y_test, groups_train, groups_test = train_test_split(X, y, iris_data['cluster'], random_state=42)
automl_settings = {
    "max_iter":5,
    "time_budget":-1,
    "metric": 'r2',
    "task": 'regression',
    "log_file_name": "error.log",
    "log_type": "all",
    "estimator_list": ['lgbm'],
    "eval_method": "cv",
    "split_type":"group",
    "groups":groups_train
}

automl.fit(X_train, y_train, **automl_settings)
```

I've investigated this issue, and found the cause. Essentially when `y_train` is a Pandas series and the `split_type` is NOT `group`, the index of y_train is reset. Applying this same `reset_index()` when `split_type='group'` fixes the issue. I will create a PR for this, shortly

Again, it's worth highlighting that if we set `split_type='uniform'` above, then the code runs without any issues.

### Expected Behavior

FLAML should be able to handle regression tasks with GroupKFold splits.

### Additional Information

FLAML 2.3.2
Python 3.10.3

**Comment by thinkall** [comment]
Thank you so much @dannycg1996  for reporting the issue. "group" is not a valid choice for regression tasks as documented [here](https://microsoft.github.io/FLAML/docs/reference/automl/automl).

Group-based splitting ensures that all data points with the same group identifier remain in the same fold for cross-validation. This is useful in situations where:
- The data is not independent (e.g., repeated measurements of a subject or data grouped by entities).
- Leakage across groups needs to be avoided in evaluation.

For regression, the task is to predict a continuous target value, and the grouping constraint may conflict with typical regression splitting practices:
If groups are not representative of the entire dataset, group-based splitting can lead to folds with imbalanced target distributions. This may degrade the model's ability to generalize.
For example, one group might predominantly cover one range of the target variable, making validation less reliable.

For regression, the goal is often to ensure the folds represent the full range of target values. Standard strategies like uniform splitting or time-based splitting ensure this more effectively.

**Comment by dannycg1996** [comment]
Hi @thinkall, thanks for the response.

I delved into the FLAML codebase as part of this issue, and got the impression that FLAML (at least originally) intended to allow `GroupKFold` to be used for regression tasks, despite the documentation. This was due to the following reasons:

- FLAML allows `GroupKFold` to be used for regression tasks. If in the example above, I remove the `train_test_split` step and just call `automl.fit(X, y)` then it runs with no issues. My issue here isn't that FLAML doesn't allow `GroupKFolds` to be used for regression tasks; my issue is that I split the dataset beforehand with `train_test_split` which means that the indices don't align if `reset_index` isn't called (as it is for other fold types).
- [This issue](https://github.com/microsoft/FLAML/issues/1254) (and response) made me think that it was the docs which were erroneous. 
-  The `decide_split_type` method on `generic_task.py` contains the following code, which suggests that the intent was to allow group-based splits for regression tasks (and explains why it does work in the cases outlined above):
  ```
          elif self.is_regression():
              assert split_type in ["auto", "uniform", "time", "group"]
              return split_type if split_type != "auto" else "uniform"
  ```
If it's really the case that `GroupKFolds` should be banned for regression tasks, the above code should be updated to explicitly forbid regression.

Also, before moving onto the next point - I will acknowledge that I shouldn't be using `train_test_split` in the original example. It doesn't invalidate this issue but in practice I'd want to consider the groups in some way when splitting the data into train and test sets!

Now in response to your points:

I acknowledge that for regression tasks, group-based splitting is less common. In most cases, users will want to keep all groups in train, validation, and test sets. But there may be circumstances in which, for regression, people want to test whether a model trained on some groups is able to predict data in other groups. It is a case-by-case situation, dependent both on the dataset and the particular problem. This issue is discussed [here] (https://stats.stackexchange.com/questions/612255/how-to-deal-with-groups-when-splitting-a-data-into-train-and-test).

In QSAR modelling (one of the use cases which I'm involved with), grouping is important for both regression and classification tasks - see [this blog post] (https://practicalcheminformatics.blogspot.com/2024/11/some-thoughts-on-splitting-chemical.html).

Apologies for the long wall of text, but I would strongly argue that FLAML should allow `GroupKFolds` to be used for regression tasks. I haven't checked all of the other AutoML tools, but I know that PyCaret does allow this use case. Thanks for reading all of that - I welcome your thoughts!
